Add unifi signal strength (dbm) attribute

SamMousa opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Unifi offers a lot of attributes in its API:
```json
{
  "site_id": "SOME HEX ID HERE",
  "assoc_time": 1616693360,
  "latest_assoc_time": 1616766090,
  "oui": "Espressi",
  "user_id": "SOME HEX ID HERE",
  "_id": "SOME HEX ID HERE",
  "mac": "MAC",
  "is_guest": false,
  "first_seen": 1603372647,
  "last_seen": 1616766090,
  "is_wired": false,
  "hostname": "xxx",
  "_uptime_by_uap": 0,
  "_last_seen_by_uap": 1616766090,
  "_is_guest_by_uap": false,
  "ap_mac": "MAC",
  "channel": 11,
  "radio": "ng",
  "radio_name": "ra0",
  "essid": "SmartEmilius",
  "bssid": "74:ac:b9:d6:c2:a9",
  "powersave_enabled": false,
  "is_11r": false,
  "network_id": "SOME HEX ID HERE",
  "user_group_id_computed": "SOME HEX ID HERE",
  "ccq": 0,
  "rssi": 63,
  "noise": -96,
  "signal": -33,
  "tx_rate": 65000,
  "rx_rate": 6000,
  "tx_power": 0,
  "idletime": 2,
  "ip": "192.168.0.59",
  "dhcpend_time": 4123168634,
  "satisfaction": 99,
  "anomalies": 0,
  "tx_mcs": 7,
  "vlan": 0,
  "radio_proto": "ng",
  "uptime": 72730,
  "tx_bytes": 539081,
  "rx_bytes": 7121573,
  "tx_packets": 9570,
  "tx_retries": 0,
  "wifi_tx_attempts": 9570,
  "rx_packets": 19044,
  "tx_bytes-r": 0,
  "rx_bytes-r": 0,
  "authorized": true,
  "qos_policy_applied": true
}
```
For identifying wifi issues before they cause a problem I want to use the `signal` attribute in a template sensor.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
Running this command reformats every file already in the repository. Instead I ran it for the 1 file I changed.

- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

SamMousa committed:
add signal strength (dbm) attribute

homeassistant commented:
Hi @SamMousa, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#48363).

Once you do that we will be able to review and accept this pull request.

Thanks!

SamMousa commented:
Unfortunately I cannot test this change since I don't have Unifi hardware at home; the location where I'd like to use it uses Home Assistant OS and I cannot easily change the hass source.

frenck commented on homeassistant/components/unifi/device_tracker.py:
```diff
@@ -40,6 +40,7 @@
     "is_guest",
     "noted",
     "qos_policy_applied",
+    "signal",
```
This should not be added as an attribute. It will cause a lot of state changes.

SamMousa replied:
What would be an alternative approach? Separate sensor(s) that you can add via the integration configuration flow?

frenck replied:
> Could also add a configuration step where you explicitly choose which attributes you want to add.

We used to call that monitored conditions, which we try to get rid off 😉 


> What would be an alternative approach? Separate sensor(s) that you can add via the integration configuration flow?

Separate sensors that are coupled with the device it belongs to, disabled by default in the entity registry on creation. That way, a user can choose to activate them.

SamMousa replied:
> Separate sensors that are coupled with the device it belongs to, disabled by default in the entity registry on creation. That way, a user can choose to activate them.

The current flow only creates them when enabled afaik: 
![image](https://user-images.githubusercontent.com/547021/112657708-f98a8780-8e52-11eb-9d51-03fe7e43515b.png)

If extending that is acceptable I'll work on it. If we need to change that flow as you mentioned by creating the sensors in a disabled state to allow individual control, I'm afraid it'll be too complicated for the time I'm able to spend on it.

Let me know if the existing flow can be extended to include a signal sensor.

Kane610 replied:
Yes, a new option is the way to go here

Kane610 replied:
And a new sensor class with test coverage, should probably be mostly copy paste work for you

frenck committed:
Bumped version to 2021.4.0b0

ludeeus committed:
Remove analytics from default_config (#48566)

bachya committed:
Fix incorrect constant import in Ambient PWS (#48574)

epenet committed:
Cleanup orphan devices in onewire integration (#48581)

* Cleanup orphan devices (https://github.com/home-assistant/core/issues/47438)

* Refactor unit testing

* Filter device entries for this config entry

* Update logging

* Cleanup check

emontnemery committed:
Move cast config flow tests to test_config_flow (#48362)

emontnemery committed:
Return config entry details for 1-step config flows (#48585)

Kane610 committed:
Don't care about DPI entries when looking for clients to be restored from UniFi (#48579)

* DPI switches shouldnt be restored, they're not part of clients to be restored

* Only care about Block and POE switch entries

youknowjack0 committed:
Fix timer.finish to cancel callback (#48549)

Timer.finish doesn't cancel the callback, which can lead to incorrect early cancellation of the timer if it is subsequently restarted. 

Bug reported here: https://community.home-assistant.io/t/timer-component-timer-stops-before-time-is-up/96038

emontnemery committed:
Include script script_execution in script and automation traces (#48576)

frenck committed:
Allow templatable service target to support scripts (#48600)

frenck committed:
Upgrade numpy to 1.20.2 (#48597)

emontnemery committed:
Include blueprint input in automation trace (#48575)

frenck committed:
Fix websocket search for related (#48603)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

bramkragten committed:
Update frontend to 20210402.0 (#48609)

balloob committed:
Clean up mobile app (#48607)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Kane610 committed:
Increase time out for http requests done in Axis integration (#48610)

balloob committed:
Bumped version to 2021.4.0b1

balloob committed:
Bumped version to 2021.4.0b2

thecode committed:
Bump aioshelly to 0.6.2 (#48620)

balloob committed:
Fix trigger template entities without a unique ID (#48631)

bramkragten committed:
Updated frontend to 20210402.1 (#48639)

balloob committed:
Support modern config for the trigger based template entity (#48635)

balloob committed:
Bumped version to 2021.4.0b3

Noltari committed:
Fix AEMET town timestamp format (#48647)

Datetime should be converted to ISO format.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

bdraco committed:
Bump aiodiscover to 1.3.3 for dhcp (#48644)

fixes #48615

bdraco committed:
Prevent config entry retry from blocking startup (#48660)

- If there are two integrations doing long retries async_block_till_done() will never be done

bdraco committed:
Only listen for zeroconf when the esphome device cannot connect (#48645)

balloob committed:
Bumped version to 2021.4.0b4

mburget committed:
Fix Raspi GPIO binary_sensor produces unreliable responses (#48170)

* Fix for issue #10498 Raspi GPIO binary_sensor produces unreliable responses ("Doorbell Scenario")

Changes overtaken from PR#31788 which was somehow never finished

* Fix for issue #10498 Raspi GPIO binary_sensor produces unreliable response. Changes taken over from PR31788 which was somehow never finished

* Remove unused code (pylint warning)

raman325 committed:
Bump zwave_js dependency to 0.23.1 (#48682)

Adminiuga committed:
Implement Ignore list for poll control configuration on Ikea devices (#48667)

Co-authored-by: Hmmbob <33529490+hmmbob@users.noreply.github.com>

balloob committed:
Fix verisure deadlock (#48691)

emontnemery committed:
Bump pychromecast to 9.1.2 (#48714)

emontnemery committed:
Flag brightness support for MQTT RGB lights (#48718)

bdraco committed:
Abort discovery for unsupported doorbird accessories (#48710)

balloob committed:
Updated frontend to 20210406.0 (#48734)

blueshiftlabs committed:
Fix infinite recursion in LazyState (#48719)

If LazyState cannot parse the attributes of its row as JSON, it prints
a message to the logger. Unfortunately, it passes `self` as a format
argument to that message, which causes its `__repr__` method to be
called, which then tries to retrieve `self.attributes` in order to
display them. This leads to an infinite recursion and a crash of the
entire core.

To fix, send the database row to be printed in the log message, rather
than the LazyState object that wraps around it.

balloob committed:
Allow reloading top-level template entities (#48733)

emontnemery committed:
Improve warnings on undefined template errors (#48713)

balloob committed:
Bumped version to 2021.4.0b5

frenck committed:
Rename hassio config entry title to Supervisor (#48748)

ludeeus committed:
Generate a seperate UUID for the analytics integration (#48742)

ludeeus committed:
Add custom integrations to analytics (#48753)

agners committed:
Use microsecond precision for datetime values on MariaDB/MySQL (#48749)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

Danielhiversen committed:
Met.no - only update data if coordinates changed (#48756)

bdraco committed:
Solve cast delaying startup when discovered devices are slow to setup (#48755)

* Solve cast delaying startup when devices are slow to setup

* Update homeassistant/components/cast/media_player.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Co-authored-by: Erik Montnemery <erik@montnemery.com>

frenck committed:
Do not activate Met.no without setting a Home coordinates (#48741)

emontnemery committed:
Fix whitespace error in cast (#48763)

bramkragten committed:
Update frontend to 20210407.0 (#48765)

emontnemery committed:
Remove login details before logging SQL errors (#48758)

frenck committed:
Bumped version to 2021.4.0b6

bramkragten committed:
Update frontend to 20210407.1 (#48778)

emontnemery committed:
Reject nan, inf from generic_thermostat sensor (#48771)

frenck committed:
Bumped version to 2021.4.0

frenck committed:
Merge pull request #48782 from home-assistant/rc

hanskroner committed:
Account for openweathermap 'dew_point' not always being present (#48826)

nzapponi committed:
Fix iCloud extra attributes (#48815)

raman325 committed:
Check all endpoints for zwave_js.climate fan mode and operating state (#48800)

* Check all endpoints for zwave_js.climate fan mode and operating state

* fix test

JohNan committed:
Add missing super call in Verisure Camera entity (#48812)

spacegaier committed:
Bump speedtest-cli to 2.1.3 (#48861)

bramkragten committed:
Correct wrong x in frontend manifest (#48865)

frenck committed:
Fix optional data payload in Prowl messaging service (#48868)

frenck committed:
Fix possibly missing changed_by in Verisure Alarm (#48867)

MartinHjelmare committed:
Fix mysensor cover closed state (#48833)

emontnemery committed:
Validate supported_color_modes for MQTT JSON light (#48836)

bramkragten committed:
Update frontend to 20210407.2 (#48888)

starkillerOG committed:
Fix motion_blinds gateway signal strength sensor (#48866)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

emontnemery committed:
Replace redacted stream recorder credentials with '****' (#48832)

balloob committed:
Bumped version to 2021.4.1

balloob committed:
Merge pull request #48896 from home-assistant/rc

milanmeu committed:
Catch expected errors and log them in rituals perfume genie (#48870)

* Add update error logging

* Move try available to else

* Remove TimeoutError

cgtobi committed:
Bump pykodi to 0.2.4 (#48913)

ludeeus committed:
Change discovery timeout from 10 to 60 (#48924)

emontnemery committed:
Extend media source URL expiry to 24h (#48912)

Ph-Wagner committed:
Extend Google Cast media source URL expiry to 24h (#48937)

* Extend media source URL expiry to 12h

closes #46280
After checking out https://github.com/home-assistant/core/pull/48912 I just think why not.

* Update homeassistant/components/cast/media_player.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

dmulcahey committed:
Bump ZHA quirks library (#48931)

ludeeus committed:
Handle exceptions when looking for new version (#48922)

cgtobi committed:
Bump pykodi to 0.2.5 (#48930)

spacegaier committed:
Fix "notify.events" trim() issue + add initial tests (#48928)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

jjlawren committed:
Fix Plex live TV handling (#48953)

balloob committed:
Bumped version to 2021.4.2

bramkragten committed:
Update frontend to 20210407.3 (#48957)

balloob committed:
Merge pull request #48954 from home-assistant/rc

balloob committed:
Fix config forwarding (#48967)

ludeeus committed:
Add TTS engines in config.components (#48939)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

jbouwh committed:
Implement percentage_step and preset_mode is not not speed fix for MQTT fan (#48951)

bdraco committed:
Prevent ping id allocation conflict with device_tracker (#48969)

* Prevent ping id allocation conflict with device_tracker

- Solves id conflict resulting unexpected home state

* Update homeassistant/components/ping/device_tracker.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

thecode committed:
Fix Shelly button device triggers (#48974)

Shutgun committed:
Bump devolo Home Control to support old websocket-client versions again (#48960)

balloob committed:
Bumped version to 2021.4.3

balloob committed:
Merge pull request #48977 from home-assistant/rc

mdz committed:
Update python-smarttub to 0.0.23 (#48978)

mdz committed:
Update python-smarttub to 0.0.23 (#48978)

bdraco committed:
Bump nexia to 0.9.6 (#48982)

- Now returns None when a humidity sensor cannot be read instead of throwing an exception

timmo001 committed:
Set Lyric hold time to use local time instead of utc (#48994)

frenck committed:
Fix use search instead of match to filter logs (#49017)

elupus committed:
Bump ha-philipsjs to 2.7.0 (#49008)

This has some improvements to not consider the TV
off due to some exceptions that is related to API
being buggy rather than off.

bdraco committed:
Bump aiohomekit to 0.2.61 (#49044)

ctalkington committed:
Resolve potential roku setup memory leaks (#49025)

* resolve potential roku setup memory leaks

* Update __init__.py

bdraco committed:
Downgrade logger message about homekit id missing (#49079)

This can happen if the TXT record is received after the PTR record and
should not generate a warning since it will get processed later

dieselrabbit committed:
Catch unknown equipment values (#49073)

* Catch unknown equipment values

* Catch unknown equipment values

* Remove warning spam.

jbouwh committed:
mqtt fan percentage to speed_range and received speed_state fix (#49060)

* percentage to speed_range and get speed state fix

* Update homeassistant/components/mqtt/fan.py

* Update homeassistant/components/mqtt/fan.py

* Update homeassistant/components/mqtt/fan.py

* Update homeassistant/components/mqtt/fan.py

Co-authored-by: J. Nick Koston <nick@koston.org>

emontnemery committed:
Fix cast options flow overwriting data (#49051)

jjlawren committed:
Check all endpoints for zwave_js.climate hvac_action (#49115)

emontnemery committed:
Quote media_source paths (#49054)

* Quote path in async_sign_path

* Address review comments, add tests

* Update tests/testing_config/media/Epic Sax Guy 10 Hours.mp4

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

unaiur committed:
Upgrade maxcube-api to 0.4.2 (#49106)

Upgrade to maxcube-api 0.4.2 to fix pending issues in HA 2021.4.x:
 - Interpret correctly S command error responses (https://github.com/home-assistant/core/issues/49075)
 - Support application timezone configuration (https://github.com/home-assistant/core/issues/49076)

thecode committed:
Fix Shelly brightness offset (#49007)

balloob committed:
Bumped version to 2021.4.4

balloob committed:
Merge pull request #49139 from home-assistant/rc

SamMousa committed:
Revert "add signal strength (dbm) attribute"

This reverts commit 7f5204a718c693bdbe4ce75c71fcc77ce08c1ad9.

SamMousa committed:
signal strength via config flow

SamMousa committed:
fixed const imports and code style

SamMousa committed:
fixed isort and flake

SamMousa committed:
isort fixes

SamMousa committed:
ran black on files affected by this PR

SamMousa committed:
first draft of tests

SamMousa commented:
Tests are passing now, but I'm not yet testing the actual reading of the signal.
to be continued.

SamMousa committed:
Merge remote-tracking branch 'upstream/master' into enh-unifi-add-signal

Kane610 commented:
Hey @SamMousa will you finalize your work here?
